File: Top_Interview_Questions/reverseLlist.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reverseListIterative(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
      return head
    }

    var last *ListNode = nil
    cur := head
    next := head.Next
    for true {
      cur.Next = last
      last = cur
      if next == nil {
        break
      }
      cur = next
      next = next.Next
    }
    return cur
}

 func reverseListRecursive(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
      return head
    }
    next := head.Next
    head.Next = nil
    root := reverseList(next)
    next.Next = head
    return root
}

File: NeetCode_Roadmap/reverse_llist.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head or !head->next) return head;

        ListNode *prev = head;
        ListNode *cur = head->next; 
        head->next = nullptr;
        while (cur) {
            ListNode *tmp = cur->next;
            cur->next = prev;
            prev = cur;
            cur = tmp;
        }
        return prev;
    }
};